### PR TITLE
Module: nest-keycloak-connect-decorators

### PR DESCRIPTION
**Pull request by meriamBenSassi**
# Nestjs-custom-decorators

A set of nestjs custom decorators.

## User decorator

A param decorator that fetches the request.user (could be added by a middleware or a guard) and add it to the params.

### Usage:

```ts
 @Get('/user')
 public getUser(@User() user: any): void {
   return user;
 }
```

## DecodeJWT

A param decorator that decodes the Json web token added to the request by by a middleware or a guard and add it to the params.

### Usage:

```ts
 @Get('/decode-jwt')
 public getDecodedJWT(@DecodeJWT() decodedJwt: any): void {
   return decodedJwt;
 }

```

### CODE REVIEW

File: packages/nest-keycloak-connect-decorators/src/decode-jwt.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// tslint:disable-next-line:variable-name
+export const DecodeJWT: () => ParameterDecorator = createParamDecorator(
+  async (_data: unknown, ctx: ExecutionContext) => {
+    const request: { accessTokenJWT: string } = ctx.switchToHttp().getRequest();
+    const jwt: string = request.accessTokenJWT;
+    let decodedJwt: object;
+    try {
+      decodedJwt = jwtDecode(jwt);
+    } catch (err) {
+      throw new Error('UNABLE_TO_DECODE_JWT');
```
Comment (ccoeurderoy) [inline packages/nest-keycloak-connect-decorators/src/decode-jwt.decorator.ts:12]: Should be an unautorized exception

File: packages/nestjs-custom-decorators/package.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  "scripts": {
+    "test": "NODE_ENV=test mocha --reporter spec --exit --require ts-node/register ./test/**/*.test.ts"
+  },
+  "bugs": {
+    "url": "https://github.com/algoan/nestjs-components/issues"
+  },
+  "dependencies": {
+    "@nestjs/common": "^7.0.7",
+    "jwt-decode": "^2.2.0",
+    "reflect-metadata": "^0.1.13"
```
Comment (ccoeurderoy) [inline packages/nestjs-custom-decorators/package.json:34]: @nestjs/common and reflect-metadata should also be in peer dependencies and dev dependencies

```suggestion
    "jwt-decode": "^2.2.0",
```

File: packages/nestjs-custom-decorators/src/decode-jwt.decorator.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+// tslint:disable-next-line:variable-name
+export const DecodeJWT: () => ParameterDecorator = createParamDecorator(
+  async (_data: unknown, ctx: ExecutionContext) => {
+    const request: { accessTokenJWT: string } = ctx.switchToHttp().getRequest();
+    const jwt: string = request.accessTokenJWT;
+    let decodedJwt: object;
+    try {
+      decodedJwt = jwtDecode(jwt);
+    } catch (err) {
+      throw new Error('UNAUTHORIZED');
```
Comment (ccoeurderoy) [inline packages/nestjs-custom-decorators/src/decode-jwt.decorator.ts:12]: Import this class from @nest/common lib

```suggestion
      throw new UnauthorizedException('UNAUTHORIZED');
```

File: packages/nestjs-custom-decorators/package.json [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    "supertest": "^4.0.2",
+    "mocha": "^7.0.0",
+    "@nestjs/core": "^7.0.7",
+    "@nestjs/platform-express": "^7.0.7",
+    "@nestjs/testing": "^7.0.7"
+  },
+  "peerDependencies": {
+    "@nestjs/common": "7.x.x",
+    "@nestjs/core": "7.x.x",
+    "@nestjs/platform-express": "7.x.x"
```
Comment (ccoeurderoy) [inline packages/nestjs-custom-decorators/package.json:50]:
```suggestion
    "@nestjs/common": ">= 7.x.x",
    "@nestjs/core": ">= 7.x.x",
    "@nestjs/platform-express": ">= 7.x.x"
```

File: packages/nestjs-custom-decorators/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# Nestjs-custom-decorators
+
+A set of nestjs custom decorators.
+
+## User decorator
+
+A param decorator that fetches the request.user (could be added by a middleware or a guard) and add it to the params.
```
Comment (g-ongenae) [inline packages/nestjs-custom-decorators/README.md:7]:
```suggestion
A param decorator that fetches the `request.user` (could be added by a middleware or a guard) and add it to the params.
```

File: packages/nestjs-custom-decorators/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+```ts
+ @Get('/user')
+ public getUser(@User() user: any): void {
+   return user;
+ }
+```
+
+## DecodeJWT
+
+A param decorator that decodes the Json web token added to the request by by a middleware or a guard and add it to the params.
```
Comment (g-ongenae) [inline packages/nestjs-custom-decorators/README.md:20]:
```suggestion
A param decorator that decodes the JSON web token added to the request by a middleware or a guard and add it to the params.
```

**Comment by meriamBenSassi** [comment]
It is ready to be merged :)

**Pull request merged**